Section Title: new component

dkallus opened the pull request:
## Description
* created new section-title component

## Context
* title sizes change based on small vs large screen

## References
New Module: Section-Title (DS6.5) https://github.com/eBay/skin/issues/524 

## Screenshots
### Large screens
<img width="1102" alt="large_1" src="https://user-images.githubusercontent.com/844822/66866605-79b67900-ef4e-11e9-9af9-2b11668ea4bc.png">
<img width="1103" alt="large_2" src="https://user-images.githubusercontent.com/844822/66866606-79b67900-ef4e-11e9-84ed-5e4bee205904.png">
<img width="1102" alt="large_3" src="https://user-images.githubusercontent.com/844822/66866607-7a4f0f80-ef4e-11e9-8fbd-1e4aaceb8079.png">
<img width="1103" alt="large_4" src="https://user-images.githubusercontent.com/844822/66866608-7a4f0f80-ef4e-11e9-9321-8f6320eb206f.png">


### Small screens
<img width="584" alt="small_1" src="https://user-images.githubusercontent.com/844822/66866618-7e7b2d00-ef4e-11e9-891a-d1938ab7604d.png">
<img width="585" alt="small_2" src="https://user-images.githubusercontent.com/844822/66866620-7e7b2d00-ef4e-11e9-9736-5ab770dc2c6a.png">
<img width="583" alt="small_3" src="https://user-images.githubusercontent.com/844822/66866621-7e7b2d00-ef4e-11e9-825b-055fa184d1b5.png">
<img width="585" alt="small_4" src="https://user-images.githubusercontent.com/844822/66866623-7f13c380-ef4e-11e9-837d-b91cba49ad62.png">

ianmcburnie commented on docs/_includes/common/section-title.html:
```diff
@@ -0,0 +1,206 @@
+    <h3>Section title</h3>
+    <p>The standard top heading is designed to support a single line on desktop and wrap only when displayed on narrow screens such as mweb or native.</p>
+
+    <div class="demo">
```
We need another `<div class="demo__inner">` child element in each demo div. This extra element helps us identify margins and margin collapse behaviour.

dkallus replied:
@ianmcburnie Is `demo_inner` needed for both documentation and tests?

dkallus replied:
Should be resolved now. I added `demo__inner` wrapper to both docs and tests

ianmcburnie commented on src/less/section-title/base/section-title.less:
```diff
@@ -0,0 +1,223 @@
+        height: 28px;
+    }
+
+    .section-title.section-title--giant {
```
Do these all need to be chained? Can we just have `.section-title--giant`? Generally we try to avoid chained selectors.

dkallus replied:
@ianmcburnie Change `.section-title.section-title--giant` to `.section-title--giant`, correct? Or are you also referring to the nested selectors inside `--giant {`?

dkallus replied:
@ianmcburnie I changed `.section-title.section-title--giant` to `.section-title--giant`. If you had other changes in mind, please let me know.

ianmcburnie commented on src/less/section-title/base/section-title.less:
```diff
@@ -0,0 +1,223 @@
+/* stylelint-disable no-descending-specificity */
```
Let's remove this if we can. As this is a new module, without any past baggage, we should aim for 0 violations of this rule.

dkallus replied:
@ianmcburnie I ran into below lint error, and I didn't know how to resolve without duplicating some styling. I tried re-ordering, but it raises a similar linting error.

Error:
```
src/less/section-title/base/section-title.less
 42:1  ✖  Expected selector ".section-title__cta > a" to come before selector                                         no-descending-specificity
          ".section-title__cta.section-title__cta--no-text > a"                                                                                
 43:1  ✖  Expected selector ".section-title__title > h2 > a" to come before selector                                  no-descending-specificity
          ".section-title__cta.section-title__cta--no-text > a"                                                                                
 52:5  ✖  Expected selector ".section-title__cta > a:visited" to come before selector ".section-title__title > h2 >   no-descending-specificity
          a:hover"
```

CSS it's complaining about:
```
.section-title__cta > a,
.section-title__title > h2 > a {
    color: @color-black;
    text-decoration: none;

    &:hover {
        color: @color-black;
        text-decoration: underline;
    }

    &:visited {
        color: @color-black;
    }
}
```

dkallus replied:
I tried this now and ran into similar issue (see below error). I think issue is with ` > h2` part. I'll separate out the stylings and remove `/* stylelint-disable no-descending-specificity */`. Duplicate styling is minimal.

```
src/less/section-title/base/section-title.less
 54:5  ✖  Expected selector ".section-title__cta > a:visited" to come before selector ".section-title__title > h2 >   no-descending-specificity
          a:hover
```

dkallus replied:
I ended up adding a new class `section-title__title-container` and moved `section-title__title` on the `h2`. Now I no longer ignore the lint rule and selectors are simpler.

agliga commented on src/less/section-title/base/section-title.less:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    > a > span:not(.section-title__cta-icon) {
+        display: none;
```
How come we have a `display: none;` on all spans except the icon.
If we are trying to hide a no-text, someone could add text without being inside a span.
Trying to understand the reasoning behind this.

Also since we already have the anchor tag nested, we can add this section inside the previous `> a` and save some keystrokes.  (although not super critical)

dkallus replied:
@agliga For this specific line, it's not needed. I removed it.

I use similar styling under small-screen media query (which I will refactor later). Reasoning is we don't want to show "See All" text next to arrow icon because small screens have limited space. As for hiding the text inside `a`, is there a way I can hide the next and not the arrow icon? Text and arrow icon are both inside the `a`.

seangates replied:
Would `.clipped` work?

dkallus replied:
@seangates I tried adding `.clipped()` mixin, but it also makes the `->` icon disappear, which we still want to show.

@agliga @seangates Another alternative is we don't handle this in CSS, but in our documentation we call out "See All" text should not be populated for small screens.

Otherwise, I think we'll need to keep `> a > span:not(.section-title__cta-icon) {` selector for small screens. In eBayUI, we would put the "See All" text inside a span, so I don't have concerns there. For users only using Skin, if they don't use span then we would still show the text.

agliga replied:
Got it, I think that's a valid reason. I do think it's good to hide stuff that's not supposed to be shown in mweb.
I do think we should also add a container class and hide it that way. Like a `section-title__see-all` or something, and then hide that specifically. That way it doesn't have this weird selector and we can just do `> a > .section-title__see-all` or whatever we name it.

dkallus replied:
I added `section-title__cta-text` class to span with text and the style selector is now just `.section-title__cta-text { display: none }` and `.section-title__cta-text { display: inline }`, depending on if it should be shown or not.

agliga commented on src/less/section-title/base/section-title.less:
```diff
@@ -0,0 +1,225 @@
+    display: inline-block;
+    margin-left: 8px;
+
+    span& {
```
I would remove these nested ones and actually put them outside. 
It might be more lines, but I do think it improves readability. 
so
```
.section-title__cta-icon {}

span.section-title__cta-icon {}

svg.section-title__cta-icon {}
```

seangates replied:
Yes, this is a pattern that makes it easier to understand the selectors. I agree.

agliga commented on src/less/section-title/base/section-title.less:
```diff
@@ -0,0 +1,225 @@
+        height: 28px;
+        width: 28px;
+    }
+}
```
These two are very similar, can we combine?
Also I think `vertical-align:middle;` might not do anything, since it works only on table's (and some inline elements). We should use `display:flex` and `align-items:center` which is the flexbox variant and works better.

seangates replied:
+1 for flexbox.

dkallus replied:
Removed `vertical-align`. It wasn't doing anything.

Also refactored styling to share width/heigh styling. For margin, I kept separate selectors since the margin-left part changes.

agliga commented on src/less/section-title/base/section-title.less:
```diff
@@ -0,0 +1,225 @@
+    display: flex;
+}
+
+@media (max-width: 600px) {
```
We should do mobile first. So the default styles should be mobile, and we should override the styles to be on dweb. 
Basically just remove the media query, add these styles to the base ones, and then the dweb one will take precedence on larger screens
(And see if we can combine any)

seangates replied:
+1 for mobile first

dkallus replied:
done

agliga commented on src/less/variables/ds6/typography-variables.less:
```diff
@@ -10,4 +10,5 @@
 @font-size-28: 1.75rem;
 
 @font-weight-bold: 700;
+@font-weight-medium: @font-weight-bold;
```
What is this for? I don't see it used here.

dkallus replied:
@agliga In section-title, I use `font-size: @font-size-medium;`, which internally usees `@font-weight-bold`. And `font-weight-bold` was not defined in DS6.

seangates commented:
Heading sizing not working? All the headings look to be he same size whether giant or small, etc.:
![image](https://user-images.githubusercontent.com/105656/66685232-19b39000-ec39-11e9-9341-7dd351add0fe.png)

dkallus commented:
@seangates Should be fixed now -- good catch :) I introduced a bug when I refactored styling to remove `/* stylelint-disable no-descending-specificity */`

ianmcburnie commented:
@dkallus Is the left margin intentional? Is that going to work as a sensible default? I wonder if we should let the page layout handle that. 

<img width="1111" alt="Screen Shot 2019-10-15 at 10 29 43 AM" src="https://user-images.githubusercontent.com/38065/66854744-c1310b00-ef36-11e9-84bd-339e559e0cbf.png">

dkallus commented:
@ianmcburnie Should I remove all margin or just `margin-left`?

Margin is intentional -- I used margins from mock. But I agree with you, most of the time user's will likely override the default margins.

ianmcburnie commented:
@dkallus Just the left margin. Margin top is a good sensible default to prevent vertically stacked modules from being right up against each other.

dkallus commented:
@ianmcburnie done :) I also updated screenshots in description

ianmcburnie commented:
Related comment: https://github.com/eBay/ebayui-core/pull/914/files/040ba1d1ff6d280eb30570e54430361a39fe9e12#r335544246

dkallus commented:
@ianmcburnie I added `aria-hidden=true` as discussed. Also pulled latest 9.4.0 changes in. Should be good to go :)
